espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/models/ActionSequence.java

```diff
@@ -41,11 +51,55 @@ public ActionSequence(W3CActions w3CActions) throws InvalidArgumentException, No
 
     @Override
     public boolean hasNext() {
-        return tickCounter < actionsByTick.size();
+        return tickCounter < ticks.size();
     }
 
     @Override
     public Tick next() {
-        return actionsByTick.get(tickCounter++);
+        return ticks.get(tickCounter++);
+    }
+
+    /**
+     * Call the dispatch algorithm defined in 17.4
+     * @param adapter
+     * @param inputStateTable
+     * @throws AppiumException
+     * @throws InterruptedException
+     * @throws ExecutionException
+     */
+    public void dispatch(W3CActionAdapter adapter, InputStateTable inputStateTable)
+            throws AppiumException, InterruptedException, ExecutionException {
+        for(Tick tick: ticks) {
+            long timeAtBeginningOfTick = System.currentTimeMillis();
+            long tickDuration = tick.calculateTickDuration();
+            List<Callable<Void>> callables = tick.dispatch(adapter, inputStateTable, tickDuration);
+
+            // 2. Wait until the following conditions are all met:
+
+            //  2.1 Wait for any pending async operations
+            if (!callables.isEmpty()) {
+                Executor executor = Executors.newFixedThreadPool(callables.size());
+                CompletionService<Void> completionService = new ExecutorCompletionService<>(executor);
+                for (Callable<Void> callable : callables) {
+                    completionService.submit(callable);
+                }
+
+                int received = 0;
+                while (received < callables.size()) {
+                    Future<Void> resultFuture = completionService.take(); //blocks if none available
```

Comment:
CompltableFuture would be a better option

Reply:
Do you have any code examples using `CompletableFuture`? I can't find anything that takes a list of callables and then translates it to a CompletableFuture

Reply:
Perhaps, in this particular case we can leave with what we have, but in general I prefer to use CompletableFuture where possible, since it can do much much more that just a generic Future.